flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java

```diff
@@ -0,0 +1,195 @@
+      this.hadoopConf = newConf;
+      return this;
+    }
+
+    public Builder env(StreamExecutionEnvironment newEnv) {
+      this.env = newEnv;
+      return this;
+    }
+
+    StreamExecutionEnvironment getEnv() {
+      return env;
+    }
+
+    RowDataTypeInfo getRowTypeInfo() {
+      return rowTypeInfo;
+    }
+
+    public FlinkInputFormat buildFormat() {
+      Preconditions.checkNotNull(tableLoader, "TableLoader should not be null");
+
+      hadoopConf = hadoopConf == null ? FlinkCatalogFactory.clusterHadoopConf() : hadoopConf;
+
+      Schema icebergSchema;
+      FileIO io;
+      EncryptionManager encryption;
+      if (table == null) {
+        // load required fields by table loader.
+        tableLoader.open(hadoopConf);
+        try (TableLoader loader = tableLoader) {
+          table = loader.loadTable();
+          icebergSchema = table.schema();
+          io = table.io();
+          encryption = table.encryption();
+        } catch (IOException e) {
+          throw new UncheckedIOException(e);
+        }
+      } else {
+        icebergSchema = table.schema();
+        io = table.io();
+        encryption = table.encryption();
+      }
+
+      if (projectedSchema != null && selectedFields != null) {
+        throw new IllegalArgumentException(
+            "Cannot using both requestedSchema and projectedFields to project");
+      }
+
+      TableSchema projectedTableSchema = projectedSchema;
+      TableSchema tableSchema = FlinkSchemaUtil.toSchema(FlinkSchemaUtil.convert(icebergSchema));
+      if (selectedFields != null) {
+        TableSchema.Builder builder = TableSchema.builder();
+        for (String field : selectedFields) {
+          TableColumn column = tableSchema.getTableColumn(field).orElseThrow(
+              () -> new IllegalArgumentException(String.format("The field(%s) can not be found in the table schema: %s",
+                  field, tableSchema)));
+          builder.field(column.getName(), column.getType());
+        }
+        projectedTableSchema = builder.build();
+      }
+
+      rowTypeInfo = RowDataTypeInfo.of((RowType) (projectedTableSchema == null ? tableSchema : projectedTableSchema)
+              .toRowDataType().getLogicalType());
+
+      Schema expectedSchema = icebergSchema;
+      if (projectedTableSchema != null) {
+        expectedSchema = FlinkSchemaUtil.convert(icebergSchema, projectedTableSchema);
+      }
+
+      return new FlinkInputFormat(tableLoader, expectedSchema, io, encryption, filterExpressions, options,
+          new SerializableConfiguration(hadoopConf));
+    }
+
+    public abstract DataStream<RowData> build();
+  }
+
+  private static final class BoundedBuilder extends Builder {
+    @Override
+    public DataStream<RowData> build() {
+      Preconditions.checkNotNull(getEnv(), "StreamExecutionEnvironment should not be null");
+      FlinkInputFormat format = buildFormat();
+      return getEnv().createInput(format, getRowTypeInfo());
```

Comment:
How is this different than an unbounded builder? I don't see anything that passes whether the stream should be bounded or unbounded. It seems like this should pass that information so that the input adapter can plan the current table scan, rather than checking for new data later.

Reply:
Not having `endSnapshot` in `ScanOptions` seems to imply that the scan is unbounded.

Reply:
I was thinking about adding a new config option like `bounded` is true or not.
After reading your comments, I think it's very good. Because streaming jobs have a high probability of setting the starting snapshot ID or timestamp, the strategy can be:
- without startSnapshotId: bounded
- with startSnapshotId and endSnapshotId: bounded
- with startSnapshotId and without endSnapshotId: unbounded
- with startSnapshotId -1 (means unbounded preceding) and without endSnapshotId: unbounded

In this way, we can have a unify builder.

Reply:
Sounds like a good way to configure this to me, except that we will want to make sure the default is reasonable. For Flink, should the default be unbounded or bounded?

Reply:
I prefer that default is bounded.
- Bounded mode is more common in attempt and startup of users.
- In unbounded mode, users often define `startSnapshotId `. But in bounded mode, `endSnapshotID` is rare. If default is unbounded, it is hard to define bounded mode.